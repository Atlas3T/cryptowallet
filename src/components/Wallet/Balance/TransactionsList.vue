<template>
  <section
    class="scroll-area"
    @touchmove="prevent"
  >
    <q-scroll-area
      ref="scrollArea"
      class="scroll-area"
    >
      <q-infinite-scroll
        :handler="loadMore"
        :offset="50"
      >
        <q-timeline
          responsive
          color="secondary"
        >
          <SingleTransaction
            v-for="transaction in filteredPaginated"
            :key="transaction.ts"
            :data="transaction"
          />
        </q-timeline>
        <q-spinner-dots
          slot="message"
          :size="40"
        />
      </q-infinite-scroll>
    </q-scroll-area>
  </section>
</template>

<script>
import SingleTransaction from './SingleTransaction.vue';

export default {
  name: 'TransactionsList',
  components: {
    SingleTransaction,
  },
  props: {
    filter: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      page: 1,
      perPage: 10,
    };
  },
  computed: {
    transactions() {
      return this.$store.state.payments.transactions;
    },
    /**
     * Filters transactions array from all/sent/received
     * @return {Array}
     */
    filtered() {
      return this.transactions.filter((transaction) => {
        if (this.filter === 'sent') return transaction.to;
        if (this.filter === 'received') return transaction.from;
        return true;
      });
    },
    /**
     * Will slice the filtered transactions array into paginated size
     * @return {Array}
     */
    filteredPaginated() {
      return this.filtered.slice(0, this.page * this.perPage);
    },
  },
  methods: {
    /**
     * Called by q-infinite-scroll component, will attempt to fetch more data
     * The second argument is a function, if said function is called with true
     * it will stop q-infinite-scroll from calling loadMore()
     * @param  {Number} index Current page, begins with 1
     * @param  {Function} done
     * @return {Boolean} true
     */
    async loadMore(index, done) {
      // if you have less data then you could fit on the page
      // it means there is no more data to be loaded
      // call done(true) to prevent subsequent attempts
      if (this.filteredPaginated.length < this.page * this.perPage) {
        done(true);
        return false;
      }

      /**
       * Included to simulate network latency
       * Without it user won't see the loading indicator
       * @param  {Number} ms Milliseconds to wait
       * @return {Function}
       */
      function sleeper(ms) {
        return function (x) {
          return new Promise(resolve => setTimeout(() => resolve(x), ms));
        };
      }

      /**
       * Wait 1500ms and the load the next page
       * await is needed for unit tests to wait for asynchronous code
       */
      await sleeper(500)().then(() => {
        this.page = this.page + 1;
        done(false);
      });

      return true;
    },
    /**
     * Stops q-pull-to-refresh from firing until TransactionsList reaches
     * the top scroll position
     * @param  {Object} event
     */
    prevent(event) {
      if (this.$refs.scrollArea.$el.childNodes[0].scrollTop > 0) {
        event.stopPropagation();
      }
    },
  },
};
</script>

<style>
.scroll-area {
  height: 100%
}
</style>
