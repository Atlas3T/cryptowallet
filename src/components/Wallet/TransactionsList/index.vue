<template>
  <section
    class="scroll-area static"
    @touchstart="touchStart"
    @touchmove="touchMove"
  >
    <q-scroll-area
      ref="scrollArea"
      class="scroll-area static"
    >
      <q-infinite-scroll
        :handler="loadMore"
        :offset="50"
      >
        <q-timeline
          responsive
          color="blueish"
        >
          <SingleTransaction
            v-for="transaction in filteredPaginated"
            :key="transaction.ts"
            :data="transaction"
          />

          <div
            v-if="filteredPaginated.length === 0"
            class="no-tx-alert"
          >
            <q-banner class="info bg-primary">
              {{ $t('emptyTransactionHistory') }}
            </q-banner>
          </div>
        </q-timeline>
        <q-spinner-dots
          slot="message"
          :size="40"
        />
      </q-infinite-scroll>
    </q-scroll-area>
  </section>
</template>

<script>
import SingleTransaction from '@/components/Wallet/SingleTransaction';
import Tx from '@/store/wallet/entities/tx';

export default {
  name: 'TransactionsList',

  components: {
    SingleTransaction,
  },

  props: {
    filter: {
      type: String,
      default: '',
    },
    wallet: {
      type: Object,
      required: true,
    },
  },

  data() {
    return {
      page: 1,
      perPage: 10,
    };
  },
  computed: {
    transactions() {
      const msTos = 1000;
      const txs = Tx.query()
        .where('wallet_id', this.wallet.id)
        .where('isChange', false)
        .get();

      txs.sort((a, b) => {
        if (this.wallet.sdk === 'Ethereum' || this.wallet.sdk === 'ERC20') {
          let b1 = new Date(b.confirmedTime * msTos);
          if (!b.confirmedTime) { b1 = new Date(b.receivedTime.toString() * msTos); }

          let a1 = new Date(a.confirmedTime * msTos);
          if (!a.confirmedTime) { a1 = new Date(a.receivedTime.toString() * msTos); }

          return b1 - a1;
        }

        if (this.wallet.sdk === 'Bitcoin') {
          return new Date(b.receivedTime * msTos) - new Date(a.receivedTime * msTos);
        }

        return 0;
      });

      return txs;
    },

    /**
     * Filters transactions array from all / sent / received
     * @return {Array}
     */
    filtered() {
      return this.transactions.filter((transaction) => {
        if (this.filter === 'sent') { return transaction.sent === true; }
        if (this.filter === 'received') { return transaction.sent === false; }
        return true;
      });
    },

    /**
     * Will slice the filtered transactions array into paginated size
     * @return {Array}
     */

    filteredPaginated() {
      return this.filtered.slice(0, this.page * this.perPage);
    },
  },

  methods: {

    /**
     * add wait to loading tx's
     * @param  {Number} ms Milliseconds to wait
     * @return {Function}
     */
    sleeper(ms) {
      return () => {
        return new Promise((resolve) => {
          return setTimeout(() => {
            return resolve();
          }, ms);
        });
      };
    },

    /**
     * Called by q-infinite-scroll component, will attempt to fetch more data
     * The second argument is a function, if said function is called with true
     * it will stop q-infinite-scroll from calling loadMore()
     * @param  {Number} index Current page, begins with 1
     * @param  {Function} done
     * @return {Boolean} true
     */
    loadMore(index, done) {
      if (this.filteredPaginated.length < this.page * this.perPage) {
        done(true);
        return false;
      }

      return this.sleeper(500)().then(() => {
        this.page = this.page + 1;
        done(false);
      });
    },

    touchStart(event) {
      this.touchStartY = event.touches[0].clientY;
    },

    touchMove(event) {
      if (event.touches[0].clientY <= this.touchStartY) {
        event.stopPropagation();
      }

      if (this.$refs.scrollArea.$el.childNodes[0].scrollTop !== 0) {
        event.stopPropagation();
      }
    },
  },
};
</script>

<style>
.scroll-area {
  height: 100%
}

.scroll-area.static {
  position: static!important;
}

.no-tx-alert {
  padding: 0.5rem;
}

.no-tx-alert .q-alert-content {
  font-size: 0.8rem;
}

.q-tab-panel {
  padding: 0;
}
</style>
