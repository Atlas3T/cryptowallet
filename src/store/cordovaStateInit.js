/*
 * this function will consume the complete Vuex store object and
 * add event listeners to Cordova's deviceready
 */
export default function (store) {
  /*
   * getDeviceContacts will attempt to get all of the devices contacts
   * as well as all of the fields available on each contact
   * if it succedes its going to mutate the Vuex store
   */
  function getDeviceContacts() {
    function onError() {
      throw new Error('getDeviceContacts(): Failed to retrieve contacts from the device.');
    }

    const options = new ContactFindOptions();
    options.multiple = true;
    const fields = ['*'];
    navigator.contacts.find(fields, (contacts) => {
      store.state.wallet.contacts = contacts;
    }, onError, options);
  }

  document.addEventListener('deviceready', () => {
    getDeviceContacts();
  }, false);
}
