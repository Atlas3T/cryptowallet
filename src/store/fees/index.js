import { Model } from '@vuex-orm/core';

export default class fees extends Model {
  static entity = 'fees';

  static primaryKey = ['code'];

  static fields() {
    return {
      code: this.attr(''),
      timestamp: this.attr(''),
      data: this.attr(''),
    };
  }
}
