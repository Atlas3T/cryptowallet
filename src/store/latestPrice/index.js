import { Model } from '@vuex-orm/core';


export default class latestPrice extends Model {
    static entity = 'latestPrice';

    static primaryKey = ['coin', 'currency']


    static fields() {
      return {
        coin: this.attr(''),
        currency: this.attr(''),
        updated: this.attr(''),
        data: this.attr(''),
      };
    }
}
