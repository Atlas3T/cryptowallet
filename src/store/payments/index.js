/* eslint-disable */


const state = {
  balance: 0,
  transactions: [],
 /* transactions: [
    {
      ts: 1539522543102,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1535527543000,
      amount: 10.00,
      to: null,
      from: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 1,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539144763020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539141943000,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1529866783020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1339527543150,
      amount: -4.97,
      to: 'Kipros',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539527543166,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539557543000,
      amount: 10.00,
      to: null,
      from: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539134763020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1531141943000,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 15439866783020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1522144743020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539181123400,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1530066783020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539517543120,
      amount: -4.97,
      to: 'Kipros',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1522141943000,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1529866553020,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1336627543150,
      amount: -4.97,
      to: 'Kipros',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539527548866,
      amount: -4.97,
      to: 'Jack Black',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1539957543000,
      amount: 10.00,
      to: null,
      from: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 15439866783330,
      amount: 8.10,
      to: null,
      from: 'Nate Diaz',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 2222144743020,
      amount: -47.97,
      to: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      from: null,
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
    {
      ts: 1566681123400,
      amount: 27.98,
      to: null,
      from: 'Cheal Sonnen',
      received: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      tx_hash: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
      status: 2,
      tx_fee: '0.000000012',
    },
  ],*/
};

export default {
  namespaced: true,
  state,
};
