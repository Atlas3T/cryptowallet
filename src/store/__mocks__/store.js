import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
