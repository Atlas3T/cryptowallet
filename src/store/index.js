/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
