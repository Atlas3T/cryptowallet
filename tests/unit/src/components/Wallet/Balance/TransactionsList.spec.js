import { shallowMount } from '@vue/test-utils';
import TransactionsList from '@/components/Wallet/Balance/TransactionsList.vue';
import { localVue, i18n, createRouter } from '@/../tests/unit/helpers/setupLocalVue.js';
import { __createMocks as createStoreMocks } from '@/../tests/store/__mocks__/store.js';

describe('TransactionsList.vue', () => {
  let storeMocks;
  let wrapper;
  let store;
  let router;

  const propsData = {
    filter: 'all',
  };

  function wrapperInit (options) {
    return shallowMount(TransactionsList, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: `/` });
    wrapper = wrapperInit({ i18n, router, localVue, store: storeMocks.store, propsData });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a section with .scroll-area class', () => {
    expect(wrapper.contains('section.scroll-area')).toBe(true);
  });

  it('calls the prevent() method on touchmove event', () => {
    const preventMock = jest.fn();
    wrapper.setMethods({prevent: preventMock});
    wrapper.find('section.scroll-area').trigger('touchmove');
    expect(preventMock).toHaveBeenCalled();
  });

  it('renders the tree of elements correctly', () => {
    expect(wrapper.contains(`section.scroll-area
                             qscrollarea-stub.scroll-area
                             qinfinitescroll-stub
                             qtimeline-stub
                             singletransaction-stub`)).toBe(true);
  });

  it('computes transactions correctly', () => {
    expect(wrapper.vm.$store.state.payments.transactions === wrapper.vm.transactions).toBe(true);
  });

  it('computes filtered correctly', () => {
    wrapper.setProps({filter: 'all'});
    expect(wrapper.vm.$store.state.payments.transactions.length === wrapper.vm.filtered.length).toBe(true);

    wrapper.setProps({filter: 'sent'});
    expect(wrapper.vm.$store.state.payments.transactions.length > wrapper.vm.filtered.length).toBe(true);
    wrapper.vm.filtered.forEach((tx) => {
      expect(tx.to).toBeTruthy();
      expect(tx.from).toBeFalsy();
    });

    wrapper.setProps({filter: 'received'});
    expect(wrapper.vm.$store.state.payments.transactions.length > wrapper.vm.filtered.length).toBe(true);
    wrapper.vm.filtered.forEach((tx) => {
      expect(tx.from).toBeTruthy();
      expect(tx.to).toBeFalsy();
    });
  });

  it('computes filteredPaginated correctly', () => {
    wrapper.setData({
      page: 1,
      perPage: 5,
    });
    expect(wrapper.vm.filteredPaginated.length).toBe(5);

    wrapper.setData({
      page: 2,
      perPage: 5,
    });
    expect(wrapper.vm.filteredPaginated.length).toBe(10);

    wrapper.setData({
      page: 2,
      perPage: 6,
    });
    expect(wrapper.vm.filteredPaginated.length).toBe(12);
  });

  it('calls done() correctly within loadMore() method', async () => {
    wrapper.setData({
      page: 1,
      perPage: 9999,
    });
    const doneMock = jest.fn();
    wrapper.vm.loadMore(null, doneMock);
    expect(doneMock).toHaveBeenLastCalledWith(true);

    wrapper.setData({
      page: 1,
      perPage: 1,
    });
    await wrapper.vm.loadMore(null, doneMock);
    expect(doneMock).toHaveBeenLastCalledWith(false);
  });

  it('calls event.stopPropagation() correctly within prevent() method', () => {
    const event = { stopPropagation: jest.fn() };
    wrapper.vm.$refs.scrollArea.$el.childNodes[0].scrollTop = 0;
    wrapper.vm.prevent(event);
    expect(event.stopPropagation).toHaveBeenCalledTimes(0);

    wrapper.vm.$refs.scrollArea.$el.childNodes[0].scrollTop = 10;
    wrapper.vm.prevent(event);
    expect(event.stopPropagation).toHaveBeenCalledTimes(1);
  });
});
