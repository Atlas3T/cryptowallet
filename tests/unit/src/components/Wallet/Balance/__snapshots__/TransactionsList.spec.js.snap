// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsList.vue renders and matches snapshot 1`] = `
<section
  class="scroll-area"
>
  <qscrollarea-stub
    class="scroll-area"
    contentactivestyle="[object Object]"
    contentstyle="[object Object]"
    delay="1000"
    thumbstyle="[object Object]"
  >
    <qinfinitescroll-stub
      handler="function () { [native code] }"
      offset="50"
    >
      <qtimeline-stub
        color="secondary"
        responsive="true"
      >
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
        <singletransaction-stub
          currency=""
          data="[object Object]"
        />
      </qtimeline-stub>
       
      <qspinnerdots-stub
        size="40"
      />
    </qinfinitescroll-stub>
  </qscrollarea-stub>
</section>
`;
